### PR TITLE
Counts number of f evaluations

### PR DESCRIPTION
**Pull request by kanav99**
This counts the number of `f` evals in `integrator.destats.nf`. Had to count this manually, so took quite time.
I don't have nice computing power right now, lets wait for travis to complete the tests.

### CODE REVIEW

File: src/perform_step/bdf_perform_step.jl [hunk 1]
```diff
@@ -153,6 +157,7 @@ function initialize!(integrator, cache::SBDFConstantCache)
   integrator.k = typeof(integrator.k)(undef, integrator.kshortsize)
   cache.du₁ = f1(uprev,p,t)
   cache.du₂ = f2(uprev,p,t)
+  integrator.destats.nf += 2
```
Comment (ChrisRackauckas) [inline src/perform_step/bdf_perform_step.jl]: I don't think we should double count the IMEX methods.

File: src/perform_step/split_perform_step.jl [hunk 2]
```diff
@@ -15,6 +16,7 @@ end
   @unpack t,dt,uprev,u,f,p = integrator
   u = @. uprev + dt*integrator.fsalfirst
   integrator.fsallast = f.f1(u,p,t+dt) + f.f2(u,p,t+dt)  # For the interpolation, needs k at the updated point
+  integrator.destats.nf += 2
```
Comment (ChrisRackauckas) [inline src/perform_step/split_perform_step.jl]: IMEX issue again

File: src/perform_step/kencarp_kvaerno_perform_step.jl [hunk 3]
```diff
@@ -283,6 +289,7 @@ end
   if typeof(integrator.f) <: SplitFunction
     integrator.k[1] = integrator.fsalfirst
     integrator.fsallast = integrator.f(u, p, t+dt)
+    integrator.destats.nf += 1
```
Comment (ChrisRackauckas) [inline src/perform_step/kencarp_kvaerno_perform_step.jl]: IMEX issue

**Comment by ChrisRackauckas** [comment]
This misses the function calls in the derivative routines.

File: src/perform_step/kencarp_kvaerno_perform_step.jl (hunk 3 above)
Comment (kanav99) [reply src/perform_step/kencarp_kvaerno_perform_step.jl]: In KenCarp3, `f2` is calculated thrice and `f1` twice. So should I increment by 3?

**Comment by kanav99** [comment]
For derivative routines, according to DiffEqDiffTools, should we do this:
```julia
tmp = length(x)
if typeof(fx) == Nothing
  tmp *= 2 # As we need to calculate both fx0 and fx
end
if eltype(df)<:Complex
  tmp *= 2 # For im and re part
end
integrator.destats.nf += tmp
```
But I doubt because `f`may assume `x` as a single element, so the statement `tmp = length(x)` might be wrong.

**Comment by ChrisRackauckas** [comment]
The Jacobian isn't handled

**Comment by ChrisRackauckas** [comment]
`tmp *= 2 # As we need to calculate both fx0 and fx`

I think forward differencing only calculates fx0 once?

**Comment by ChrisRackauckas** [comment]
>if eltype(df)<:Complex

complex differentiation doesn't occur when eltype is complex.

**Comment by kanav99** [comment]
For the complex issue, I was trying to count this evaluation: https://github.com/JuliaDiffEq/DiffEqDiffTools.jl/blob/master/src/gradients.jl#L183 We have calculated f(x) twice in case df is complex. Am I missing something?

**Comment by kanav99** [comment]
This commit fails the test as algorithms like `NorsettEuler, ETDRK2, ETDRK3, ETDRK4, HochOst4` don't have the field diff_type. I am not sure how to proceed from here

**Comment by ChrisRackauckas** [comment]
>This commit fails the test as algorithms like NorsettEuler, ETDRK2, ETDRK3, ETDRK4, HochOst4 don't have the field diff_type. I am not sure how to proceed from here. I will investigate a bit more.

Those are exponential integrators. You can just count the `f2` on those.

**Comment by ChrisRackauckas** [comment]
>For the complex issue, I was trying to count this evaluation: https://github.com/JuliaDiffEq/DiffEqDiffTools.jl/blob/master/src/gradients.jl#L183 We have calculated f(x) twice in case df is complex. Am I missing something?

@dextorious I don't think it needs both `f` evaluations, right?

**Comment by ChrisRackauckas** [comment]
I realized that we should add some really good tests for this. You can create a function which has a global ref `counter = Ref(0)` and then in the ODE function do `counter[] += 1`. That will give exactly how many function calls are done. Then you can run the algorithm and compare this to DEStats. Do this on the main algorithms from each class (excluding exponential integrator for now, since they will need some weird different counter), so I would test `BS3`, `Tsit5`, `Vern7/8/9`, `KenCarp4`, `Rosenbrock23`, `Rodas5`, and switching up the differentiation styles on them. Mark the ones that are broken with `@test_broken` and we can merge, and continue to fix more of them in further PRs.

**Comment by kanav99** [comment]
Okay this sounds like a plan. I was actually calculating number for automatic differentiation which turned out to be one only. Only this case of autodiff is a bit peculiar. I will quickly write the tests and push them first.

**Comment by kanav99** [comment]
In this commit: 
1. In autodiff, I have incremented by 1 only, though each call of `f` is a bit expensive
2. `autodiff=false` case doesn't give accurate results.
Observation: On changing `length(u0)` the deficit from actual value doesn't change. For eg: Currently length(u0) = 2 in the tests the deficit is 28 (255-227) but on `length(u0) = 1` deficit is still same. So bug is not in the `derivative` or `jacobian`.

**Comment by ChrisRackauckas** [comment]
This is very strong start at least. Thanks!

**Pull request merged**